### ISSUE TITLE
GDL Criterion error

### ISSUE THREAD

**Issue by omair-kg** [coupriec/VideoPredictionICLR2016#7]
Hi,

I'm having problems using the GDL criterion. The error stack states
../torch/install/share/lua/5.1/nn/CSubTable.lua:11: bad argument #2 to 'add' (sizes do not match at /tmp/luarocks_cutorch-scm-1-7235/cutorch/lib/THC/generated/../generic/THCTensorMathPointwise.cu:216)
stack traceback:
        [C]: in function 'add'
        ../torch/install/share/lua/5.1/nn/CSubTable.lua:11: in function 'func'
        ../torch/install/share/lua/5.1/nngraph/gmodule.lua:346: in function 'neteval'
        ../torch/install/share/lua/5.1/nngraph/gmodule.lua:381: in function 'updateOutput'
        ./GDLCrit.lua:40: in function 'forward'

The size of the output does indeed change (it's smaller by one pixel). Before the target label and the output of the network is forwarded through the gdl loss they are both the same size. I haven't been able to figure out where the size change occurs. Any help would be really appreciated.

**Comment by MichaelMathieu** [comment]
Hi,

Internally inside the GDL, the images are indeed reduced by one pixel (because we compute the spatial gradients using finite differences). It happens in the nn.SpatialZeroPadding modules (padding with a negative amount of pixels is cropping).
I don't understand the exact nature of the error, especially since the file GDLCrit.lua doesn't correspond to a file in the repo (so what is on line 40?).
If you cannot figure out the error, please paste a full, runnable code that can produce the error.

**Comment by omair-kg** [comment]
Hi,

I copied the GDL criterion into its own file. Line 40 corresponds to:
function GDL:updateOutput(input, target)
    self.netoutput = self.net:updateOutput{input, target}

In the repo this is line 329 in the file get_model.lua

**Comment by omair-kg** [comment]
Hi @MichaelMathieu 

This is the minimum working code for generating the error:

```
require 'torch'
require 'nn'
require 'optim'
require 'GDLCrit'

criterionGDL = nn.GDLCriterion(1)
x = torch.Tensor(1,1,38,64)
y = torch.Tensor(1,1,38,64)
errGDL = criterionGDL:forward(y,x)
print(#errGDL)
df_do_gdl = criterionGDL:backward(y,x)
print(#df_do_gdl)
```
The GDLCrit.lua file is : 
```
-- Written by Authors: Michael Mathieu, Camille Couprie, July 2016
--ICLR 2016 http://cs.nyu.edu/~mathieu/iclr2016.html
-- Gradient Difference Loss
require 'nnx'
require 'nngraph'
GDL, gdlparent = torch.class('nn.GDLCriterion', 'nn.Criterion')

function GDL:__init(alpha)
    gdlparent:__init(self)
    self.alpha = alpha or 1
    assert(alpha == 1) --for now
    local Y = nn.Identity()()
    local Yhat = nn.Identity()()
    local Yi1 = nn.SpatialZeroPadding(0,0,0,-1)(Y)
    local Yj1 = nn.SpatialZeroPadding(0,0,-1,0)(Y)
    local Yi2 = nn.SpatialZeroPadding(0,-1,0,0)(Y)
    local Yj2 = nn.SpatialZeroPadding(-1,0,0,0)(Y)
    local Yhati1 = nn.SpatialZeroPadding(0,0,0,-1)(Yhat)
    local Yhatj1 = nn.SpatialZeroPadding(0,0,-1,0)(Yhat)
    local Yhati2 = nn.SpatialZeroPadding(0,-1,0,0)(Yhat)
    local Yhatj2 = nn.SpatialZeroPadding(-1,0,0,0)(Yhat)
    local term1 = nn.Abs()(nn.CSubTable(){Yi2, Yi1})
    local term2 = nn.Abs()(nn.CSubTable(){Yhati2,  Yhati1})
    local term3 = nn.Abs()(nn.CSubTable(){Yj2, Yj1})
    local term4 = nn.Abs()(nn.CSubTable(){Yhatj2, Yhatj1})
    local term12 = nn.CSubTable(){term1, term2}
    local term34 = nn.CSubTable(){term3, term4}
    self.net = nn.gModule({Yhat, Y}, {term12, term34})
    --self.net:cuda()
    self.crit = nn.ParallelCriterion()
    self.crit:add(nn.AbsCriterion())
    self.crit:add(nn.AbsCriterion())
    --self.crit:cuda()
    self.target1 = torch.Tensor()
    self.target2 = torch.Tensor()
end

function GDL:updateOutput(input, target)
    self.netoutput = self.net:updateOutput{input, target}
    self.target1:resizeAs(self.netoutput[1]):zero()
    self.target2:resizeAs(self.netoutput[2]):zero()
    self.target = {self.target1, self.target2}
    self.loss = self.crit:updateOutput(self.netoutput, self.target)
    return self.loss
end

function GDL:updateGradInput(input, target)
    local gradInput =
        self.crit:updateGradInput(self.netoutput, self.target)
    self.gradInput =
        self.net:updateGradInput({input, target}, gradInput)[1]
    return self.gradInput
end
```

Where the line corresponding to the error is: 
`self.netoutput = self.net:updateOutput{input, target}`

The error is 

> .../torch/install/share/lua/5.1/nn/CSubTable.lua:11: inconsistent tensor size at /tmp/luarocks_torch-scm-1-1578/torch7/lib/TH/generic/THTensorMath.c:601
> stack traceback:
> 	[C]: in function 'add'
> 	.../torch/install/share/lua/5.1/nn/CSubTable.lua:11: in function 'func'
> 	.../torch/install/share/lua/5.1/nngraph/gmodule.lua:345: in function 'neteval'
> 	.../torch/install/share/lua/5.1/nngraph/gmodule.lua:380: in function 'updateOutput'
> 	./GDLCrit.lua:39: in function 'forward'

Thanks in anticipation.

**Comment by shahabty** [comment]
Hello, 
Do you know how to handle dimension mismatching error?
I've implemented the pytorch version of this GDL loss. But I faced exactly the same error.

**Comment by javierselva** [comment]
HI!
I am actually having some trouble finding where in the code is the GDL loss being used. I see the declaration of the class on get_model.lua but it is not called from anywhere. And while I see the weights for the L2 loss and Adversarial loss in the parameters definition at the beginning of train_iclr_model.lua I do not see any for GDL. I wanted to modify the weight for this loss in order to see how much effect does it have on the prediction's quality.

Should I implement it myself? Or am I missing something?

Thanks in advance!
